### ISSUE TITLE
Gigabyte Waterforce 240 (Non X variant)

### ISSUE THREAD

**Issue by 7h3ju57** [aleksamagicka/waterforce-hwmon#1]
Found you through this thread https://github.com/liquidctl/liquidctl/issues/167
Would it be possible to add support for the Waterforce 240 (Non X Variant)?

Bus 003 Device 002: ID 1044:7a51 Chu Yuen Enterprise Co., Ltd GIGABYTE CPU Cooler
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               1.10
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1044 Chu Yuen Enterprise Co., Ltd
  idProduct          0x7a51 
  bcdDevice            1.00
  iManufacturer           1 ROCCAT
  iProduct                2 GIGABYTE CPU Cooler
  iSerial                 3 000000000000
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0062
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      24
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           2
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      34
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x04  EP 4 OUT
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
      Warning: Descriptor too short
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         2
          bDescriptorType        34 Report
          wDescriptorLength      47
          bDescriptorType         0 (null)
          wDescriptorLength       0
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               4
can't get debug descriptor: Resource temporarily unavailable
Device Status:     0x0000
  (Bus Powered)

**Comment by aleksamagicka** [comment]
Looks similar, we can always try! Please try out the [non-x-id](https://github.com/aleksamagicka/waterforce-hwmon/tree/non-x-id) branch and report back.

**Comment by 7h3ju57** [comment]
Thanks for looking into it.
It compiled!
shows up three times in /sys/class/hwmon/*/name

sensors output
waterforce-hid-3-2
Adapter: HID adapter
Fan speed:                   N/A
Pump speed:                  N/A
Coolant temp:                N/A  
User provided CPU temp:      N/A  

waterforce-hid-3-3
Adapter: HID adapter
Fan speed:              32616 RPM
Pump speed:                0 RPM
Coolant temp:           -1055612.9°C  
User provided CPU temp:      N/A  

waterforce-hid-3-1
Adapter: HID adapter
Fan speed:                   N/A
Pump speed:                  N/A
Coolant temp:                N/A  
User provided CPU temp:      N/A

**Comment by 7h3ju57** [comment]
if i try to remove it run sensors-detect, the system no longer responds and needs a hard reset

**Comment by aleksamagicka** [comment]
Thanks. Looks like only one of the listed devices is the true one and it also needs work. I should be getting an GIGABYTE AORUS WATERFORCE 280 (also non X) tomorrow, so we'll be on the same page.

**Comment by 7h3ju57** [comment]
heres some output from syslog if you need it
let me know if there's anything else i can do to help.
Thanks again!

May 16 14:36:28 phoenix kernel: [   61.193838] waterforce 0003:1044:7A51.0001: hidraw0: USB HID v1.11 Mouse [ROCCAT GIGABYTE CPU Cooler] on usb-0000:06:00.3-1/input0
May 16 14:36:28 phoenix kernel: [   61.253876] waterforce 0003:1044:7A51.0001: fw version request failed with -38
May 16 14:36:28 phoenix kernel: [   61.254014] waterforce: probe of 0003:1044:7A51.0001 failed with error -38
May 16 14:36:28 phoenix kernel: [   61.254297] waterforce 0003:1044:7A51.0002: hidraw0: USB HID v1.11 Device [ROCCAT GIGABYTE CPU Cooler] on usb-0000:06:00.3-1/input1
May 16 14:36:30 phoenix kernel: [   63.416311] waterforce 0003:1044:7A51.0002: fw version request failed with -61
May 16 14:36:30 phoenix kernel: [   63.416585] waterforce: probe of 0003:1044:7A51.0002 failed with error -61
May 16 14:36:30 phoenix kernel: [   63.416796] waterforce 0003:1044:7A51.0003: hidraw0: USB HID v1.11 Device [ROCCAT GIGABYTE CPU Cooler] on usb-0000:06:00.3-1/input2
May 16 14:36:30 phoenix kernel: [   63.476281] waterforce 0003:1044:7A51.0003: fw version request failed with -38
May 16 14:36:30 phoenix kernel: [   63.476395] waterforce: probe of 0003:1044:7A51.0003 failed with error -38
May 16 14:36:34 phoenix kernel: [   66.727232] BUG: kernel NULL pointer dereference, address: 0000000000000000
May 16 14:36:34 phoenix kernel: [   66.727237] #PF: supervisor write access in kernel mode
May 16 14:36:34 phoenix kernel: [   66.727239] #PF: error_code(0x0002) - not-present page
May 16 14:36:34 phoenix kernel: [   66.727240] PGD 0 P4D 0 
May 16 14:36:34 phoenix kernel: [   66.727242] Oops: 0002 [#1] SMP NOPTI
May 16 14:36:34 phoenix kernel: [   66.727244] CPU: 5 PID: 8451 Comm: sensors Tainted: P           OE     5.15.0-72-generic #79-Ubuntu
May 16 14:36:34 phoenix kernel: [   66.727247] Hardware name: System manufacturer System Product Name/TUF GAMING X570-PLUS (WI-FI), BIOS 4408 10/28/2022
May 16 14:36:34 phoenix kernel: [   66.727249] RIP: 0010:__mutex_lock.constprop.0+0x16b/0x490
May 16 14:36:34 phoenix kernel: [   66.727253] Code: fb 01 00 48 89 44 24 20 49 8d 46 10 4c 8d 6c 24 10 48 89 c3 48 89 04 24 49 8b 46 18 48 89 5c 24 10 4d 89 6e 18 48 89 44 24 18 <4c> 89 28 4d 39 6e 10 0f 84 b6 02 00 00 65 48 8b 1c 25 c0 fb 01 00
May 16 14:36:34 phoenix kernel: [   66.727256] RSP: 0018:ffffbc7c446cfbb0 EFLAGS: 00010293
May 16 14:36:34 phoenix kernel: [   66.727257] RAX: 0000000000000000 RBX: ffff9af957246950 RCX: ffff9af951de3280
May 16 14:36:34 phoenix kernel: [   66.727259] RDX: 0000000000000001 RSI: ffff9af97a854188 RDI: ffff9af957246948
May 16 14:36:34 phoenix kernel: [   66.727260] RBP: ffffbc7c446cfc20 R08: ffffbc7c446cfcc0 R09: ffff9af940d7d000
May 16 14:36:34 phoenix kernel: [   66.727261] R10: 000000000000000b R11: 0000000000000000 R12: ffff9af957246940
May 16 14:36:34 phoenix kernel: [   66.727262] R13: ffffbc7c446cfbc0 R14: ffff9af957246940 R15: ffff9af957246948
May 16 14:36:34 phoenix kernel: [   66.727264] FS:  00007f13f7554740(0000) GS:ffff9b084eb40000(0000) knlGS:0000000000000000
May 16 14:36:34 phoenix kernel: [   66.727265] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
May 16 14:36:34 phoenix kernel: [   66.727267] CR2: 0000000000000000 CR3: 000000024e8b2000 CR4: 0000000000750ee0
May 16 14:36:34 phoenix kernel: [   66.727268] PKRU: 55555554
May 16 14:36:34 phoenix kernel: [   66.727270] Call Trace:
May 16 14:36:34 phoenix kernel: [   66.727271]  <TASK>
May 16 14:36:34 phoenix kernel: [   66.727276]  __mutex_lock_slowpath+0x13/0x20
May 16 14:36:34 phoenix kernel: [   66.727278]  mutex_lock+0x38/0x50
May 16 14:36:34 phoenix kernel: [   66.727280]  waterforce_write_expanded+0x24/0x150 [waterforce]
May 16 14:36:34 phoenix kernel: [   66.727283]  waterforce_read+0x7f/0x170 [waterforce]
May 16 14:36:34 phoenix kernel: [   66.727285]  hwmon_attr_show+0x3f/0xd0
May 16 14:36:34 phoenix kernel: [   66.727287]  dev_attr_show+0x1d/0x50
May 16 14:36:34 phoenix kernel: [   66.727289]  sysfs_kf_seq_show+0xa5/0x100
May 16 14:36:34 phoenix kernel: [   66.727292]  kernfs_seq_show+0x27/0x30
May 16 14:36:34 phoenix kernel: [   66.727294]  seq_read_iter+0x124/0x4b0
May 16 14:36:34 phoenix kernel: [   66.727297]  ? _copy_to_user+0x20/0x30
May 16 14:36:34 phoenix kernel: [   66.727300]  kernfs_fop_read_iter+0x30/0x40
May 16 14:36:34 phoenix kernel: [   66.727302]  new_sync_read+0x10d/0x190
May 16 14:36:34 phoenix kernel: [   66.727304]  vfs_read+0x103/0x1a0
May 16 14:36:34 phoenix kernel: [   66.727306]  ksys_read+0x67/0xf0
May 16 14:36:34 phoenix kernel: [   66.727307]  __x64_sys_read+0x19/0x20
May 16 14:36:34 phoenix kernel: [   66.727309]  do_syscall_64+0x5c/0xc0
May 16 14:36:34 phoenix kernel: [   66.727312]  ? do_syscall_64+0x69/0xc0
May 16 14:36:34 phoenix kernel: [   66.727313]  ? exit_to_user_mode_prepare+0x37/0xb0
May 16 14:36:34 phoenix kernel: [   66.727316]  ? syscall_exit_to_user_mode+0x27/0x50
May 16 14:36:34 phoenix kernel: [   66.727318]  ? do_syscall_64+0x69/0xc0
May 16 14:36:34 phoenix kernel: [   66.727320]  entry_SYSCALL_64_after_hwframe+0x61/0xcb
May 16 14:36:34 phoenix kernel: [   66.727322] RIP: 0033:0x7f13f7752992
May 16 14:36:34 phoenix kernel: [   66.727323] Code: c0 e9 b2 fe ff ff 50 48 8d 3d fa b2 0c 00 e8 c5 1d 02 00 0f 1f 44 00 00 f3 0f 1e fa 64 8b 04 25 18 00 00 00 85 c0 75 10 0f 05 <48> 3d 00 f0 ff ff 77 56 c3 0f 1f 44 00 00 48 83 ec 28 48 89 54 24
May 16 14:36:34 phoenix kernel: [   66.727326] RSP: 002b:00007fff4ecd8d48 EFLAGS: 00000246 ORIG_RAX: 0000000000000000
May 16 14:36:34 phoenix kernel: [   66.727328] RAX: ffffffffffffffda RBX: 0000557dc89ac370 RCX: 00007f13f7752992
May 16 14:36:34 phoenix kernel: [   66.727329] RDX: 0000000000001000 RSI: 0000557dc89a5820 RDI: 0000000000000003
May 16 14:36:34 phoenix kernel: [   66.727330] RBP: 00007f13f7854600 R08: 0000000000000003 R09: 0000557dc89a5820
May 16 14:36:34 phoenix kernel: [   66.727331] R10: 00007f13f78583e0 R11: 0000000000000246 R12: 00007f13f7858580
May 16 14:36:34 phoenix kernel: [   66.727332] R13: 0000000000000d68 R14: 00007f13f7853a00 R15: 0000000000000d68
May 16 14:36:34 phoenix kernel: [   66.727334]  </TASK>
May 16 14:36:34 phoenix kernel: [   66.727335] Modules linked in: waterforce(OE) nfsv3 nfs_acl rpcsec_gss_krb5 auth_rpcgss nfsv4 nfs lockd grace fscache netfs veth xt_nat nf_conntrack_netlink xfrm_user xfrm_algo xt_addrtype br_netfilter xt_CHECKSUM xt_MASQUERADE xt_conntrack ipt_REJECT nf_reject_ipv4 xt_tcpudp nft_compat nft_chain_nat nf_nat nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 nft_counter nf_tables nfnetlink overlay bridge stp llc binfmt_misc sunrpc intel_rapl_msr intel_rapl_common edac_mce_amd nls_iso8859_1 kvm_amd snd_hda_codec_realtek kvm snd_hda_codec_generic ledtrig_audio snd_hda_codec_hdmi btusb rapl btrtl btbcm snd_hda_intel btintel snd_intel_dspcfg snd_intel_sdw_acpi bluetooth eeepc_wmi wmi_bmof snd_hda_codec joydev snd_hda_core zenpower(OE) ecdh_generic snd_hwdep ecc ccp snd_pcm cp210x usbserial snd_timer snd soundcore nvidia_uvm(POE) mac_hid sch_fq_codel dm_multipath scsi_dh_rdac scsi_dh_emc scsi_dh_alua msr ramoops pstore_blk pstore_zone efi_pstore reed_solomon ip_tables x_tables autofs4 btrfs
May 16 14:36:34 phoenix kernel: [   66.727366]  blake2b_generic zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear nvidia_drm(POE) nvidia_modeset(POE) hid_generic usbhid hid nvidia(POE) amdgpu iommu_v2 gpu_sched i2c_algo_bit drm_ttm_helper ttm mfd_aaeon drm_kms_helper asus_wmi sparse_keymap video platform_profile syscopyarea sysfillrect sysimgblt crct10dif_pclmul crc32_pclmul fb_sys_fops ghash_clmulni_intel cec nvme r8169 rc_core aesni_intel ahci xhci_pci crypto_simd cryptd drm i2c_piix4 nvme_core libahci realtek xhci_pci_renesas wmi
May 16 14:36:34 phoenix kernel: [   66.727397] CR2: 0000000000000000
May 16 14:36:34 phoenix kernel: [   66.727398] ---[ end trace 316279b6553a4caa ]---

**Comment by aleksamagicka** [comment]
Thanks, I think I know what causes that... But to fully fix I'll need to investigate once I get the device.

**Comment by aleksamagicka** [comment]
I don't know if Gigabyte does _any_ QC on their software. I've had at least 5 BSODs in a matter of hours since I downloaded their software (which, by the way, does not recognize this AIO - only Aorus Engine does, which is equally disastrous.) They are extremely annoying and interrupt my work.

So I haven't made any significant progress, although some parts do overlap with the X-series, which is useful. I've rigged something up on my Linux install, but it's returning giberrish. Perhaps it requires an init before doing anything, I see the software creating a flurry of requests once opened. Once I figure something out I'll post here.

**Comment by aleksamagicka** [comment]
I've pushed a commit to the [non-x-id](https://github.com/aleksamagicka/waterforce-hwmon/tree/non-x-id) branch. Pump and fan speed should now be read correctly. Please test if possible, thanks.

**Comment by 7h3ju57** [comment]
Only one instance showing up this time but N/A across the board.

waterforce-hid-3-3
Adapter: HID adapter
Fan speed:                   N/A
Pump speed:                  N/A
User provided CPU temp:      N/A


May 22 17:51:47 phoenix kernel: [213126.515444] waterforce 0003:1044:7A51.0003: hidraw0: USB HI
D v1.11 Device [ROCCAT GIGABYTE CPU Cooler] on usb-0000:06:00.3-1/input2
May 22 17:51:47 phoenix kernel: [213126.577917] waterforce 0003:1044:7A51.0003: fw version requ
est failed with -19


Edit
And system is now unresponsive lol

**Comment by aleksamagicka** [comment]
At first I just thought that I was missing some init commands. Turns out that the "flurry of requests" from earlier is actually the software sending the whole firmware to the device... The first thing that tipped me off is that one of the reports had `KERNEL32.DLL` in it, then I found the E7 command at the end ([`WritingFlash - 0xE7`](https://github.com/liquidctl/liquidctl/issues/167#issuecomment-1328582258)). Then the normal requests followed.

Unfortunately, I don't think there's anything we can do here. There are three firmware files, called `0.bin`, `90.bin` and `180.bin`, referring to the display orientation most likely, each one is 10MB and most importantly, we don't have the rights to distribute them anywhere (plus the problem of firmware updates). The reason it worked for me is because I was rebooting from Windows to Linux and the firmware stayed in the device RAM.
